homeassistant/components/tplink/light.py

```diff
@@ -61,19 +63,34 @@
 async def async_setup_entry(hass: HomeAssistantType, config_entry, async_add_entities):
     """Set up lights."""
     devices = hass.data[TPLINK_DOMAIN][CONF_LIGHT]
-    entities = []
-
-    await hass.async_add_executor_job(get_devices_sysinfo, devices)
-    for device in devices:
-        entities.append(TPLinkSmartBulb(device))
-
-    async_add_entities(entities, update_before_add=True)
+    entities_ready = []
+    entities_unavailable = []
+
+    entities_ready, entities_unavailable = await hass.async_add_executor_job(
+        get_devices_sysinfo, devices, TPLinkSmartBulb
+    )
+    if entities_ready:
+        async_add_entities(entities_ready, update_before_add=True)
+
+    async def async_retry_setup_entry(_):
```

Comment:
Maybe we can just raise `PlatformNotReady` instead? That will retry the platform set up later with a back off timer. If we keep track of the entities we add at every attempt there will be no duplication.

Reply:
I agree. That should make this a lot cleaner.

Reply:
Is there an example of how to track the unavailable devices? Will that mean the available devices will still be added and usable using PlatformNotReady?

Reply:
The exception we raise is just a way to tell the entity platform helper to try the platform setup again later. It's up to us to add the correct entities.

So we have to check that the device is available first, then create the entity, if we haven't already added the corresponding entity for that device. We can do this by storing the unique id in a set. Add all ok entities. Then if there are devices that are not available, raise `PlatformNotReady`.

Reply:
That logic makes sense, I am just not sure do I need to / can I store just the ones that haven't been added so that when it re-runs it only runs against those, or is it better to just check that the devices haven't already been added before attempting to add again.

Reply:
Both ways are ok, I think, as long as we can guarantee not to add duplicate entities.

Reply:
Other than moving the logic into the common so it isn't as redundant do you think what I have is ok or I should still switch to the PlatfromNotReady? I think what is confusing me about the PlatformNotReady is it looks like it is currently used when HA / can't connect to an actual platform, not just an individual device. Maybe it doesn't matter but the exception name sounds misleading in this case.

Reply:
I think using `PlatformNotReady` will clean up the code.

Reply:
You can keep track of the unavailable entities in `hass.data['tplink_unavailable_light']`

Reply:
Just set it to a list. As long as the key starts with the integration domain, you can do as you wish.

Reply:
got it, thanks guys, will make the update as suggested